### PR TITLE
Print `initialized` stanza to each file

### PR DESCRIPTION
**Pull request by gza060625**
When using `-Xverbosegclog`, prints `Initialized` stanza at the beginning of each file.
In order to achieve this, following changes have been made:
- Relocate `formatAndOutput(),formatAndOutputV()` to `MM_VerboseBuffer` from `MM_VerboseWriterChain`
- In `MM_VerboseHandlerOutput`:
   - Modified `writeVmArgs` to take a `MM_VerboseBuffer`
   - Factored out lines for printing `Initialized` stanza in `handleInitialized`
   - Created `outputInitializedStanza` dedicated to output `Initialized` stanza, which also takes a `MM_VerboseBuffer`
   - In `outputInitializedStanza`, replaced `handleInitializedInnerStanzas` with `outputInitializedInnerStanza` 
- In `VerboseHandlerOutput`:
   - Modified `writeVmArgs` to takes a `MM_VerboseBuffer`
   - Implement virtual method `outputInitializedInnerStanza()`
- Implement `MM_VerboseManager::getVerboseHandlerOutput()`
- In `VerboseWriterChain`:
   - Delete `formatAndOutputV()`
   - Redirect `formatAndOutput()` to `MM_VerboseBuffer`
   - Implement `getBuffer()`
- In `MM_VerboseWriterFileLoggingBuffered` and `VerboseWriterFileLoggingSynchronous`:
   - Print `Initialized` stanza on the second file opens (First file is handled by `TRIGGER_J9HOOK_MM_OMR_INITIALIZED`)
- In `MM_VerboseWriterFileLogging::endOfCycle`, opens a file right after last file closes to make sure a new `Initialized` stanza is printed in `openFile(...)`.
- Changes to generalize handleIniaitzed to work on any provided buffer rather than writer chain specific buffer. This is done to bypass the writer chain with a new buffer during openFile to print stanza specifically for the file being opened.

### CODE REVIEW

**Review by RSalman — CHANGES_REQUESTED** [review]

File: gc/base/ModronAssertions.h
Comment (gza060625) [inline gc/base/ModronAssertions.h:35]: @RSalman I am not sure about these lines. But I have to add them to remove redefinition errors.

Comment (RSalman) [reply gc/base/ModronAssertions.h:35]: This shouldn't be necessary, it doesn't look right. I'm not sure why, but there must be an underlying def issue in the changes.

Comment (gza060625) [reply gc/base/ModronAssertions.h]: @RSalman 
I agree this is one aggressive approach; it may jeopardize the the point to have these variables as guards. 
The definition issue arises from `VerboseHandlerJava.cpp` which includes `VerboseBuffer.hpp` in recent changes.
`VerboseHandlerJava.cpp=>VerboseBuffer.hpp=>ut_j9vgc`
`VerboseHandlerJava.cpp=>FinalizeListManager=>ut_j9mm`

Both `ut_j9vgc` and `ut_j9mm` define above variables, thus the redefinition. Since the issue is in the `VerboseHandlerJava.cpp`, they should be undefined in `VerboseHandlerJava.cpp` as well. Please check my new commit.

https://github.com/eclipse/omr/pull/5607/commits/6cb15bd9b4817c0e7d1943b8cf11b0fc3c290e04
https://github.com/eclipse/openj9/pull/10879/commits/15c8bbf7c36ce20ab639701a1fb7e39dd86c582e

File: gc/verbose/VerboseWriterFileLoggingSynchronous.cpp
Comment (RSalman) [reply gc/verbose/VerboseWriterFileLoggingSynchronous.cpp:124]: I know you had introduced a new class flag before I suggested using `isAcitve` for guarding the print. But let's try not to use `isActive`, there's no clear relationship between isActive and printing the init stanza. We're just piggy backing off the fact that writers are activated after GC initialized event is handled (i.e after first file init is printed by event) which is not clear by looking at this.

Comment (RSalman) [reply gc/verbose/VerboseWriterFileLoggingSynchronous.cpp:124]: There's 3 approaches we can take,

1. Move this init print block down to when openFile is called in `outputString` 
2. Introduce new bool param to openFile, which is set to true when the opening the file that requires the init stanza printed (which is all call sites except the openFile done during initialize) 
3. Class flag as you had before

Comment (RSalman) [reply gc/verbose/VerboseWriterFileLoggingSynchronous.cpp:124]: Class flag is a bit much as it'll always be true after first file is opened. Out of those approches, I like 2.

**Review by RSalman — CHANGES_REQUESTED** [review]

File: gc/base/ModronAssertions.h
Comment (RSalman) [reply gc/base/ModronAssertions.h]: ah okay, that makes sense. It looks okay and it should be fine, although, I don't know if it's "the right approach" or if we should work around it. The hesitation comes from reading existing comments saying _" OMRTODO temporary workaround to allow .... "_ in some places were something similar  is done. Leave it for now and I'll try to take a look.

Comment (RSalman) [reply gc/base/ModronAssertions.h]: > VerboseHandlerJava.cpp=>FinalizeListManager=>ut_j9mm

I don't see `ut_j9mm` being included in `FinalizeListManager.hpp`, am I missing something?

Also, if you haven't already, try switching order of includes. `VerboseBuffer.hpp` already undef these, try putting `VerboseBuffer.hpp` last, after `FinalizeListManager`.

Comment (gza060625) [reply gc/base/ModronAssertions.h]: My Bad, the complete sequence should be:
`VerboseHandlerJava.cpp=>FinalizeListManager=>ObjectAccessBarrier=>ModronAssertions=>ut_j9mm`
```
In file included from /root/code/openj9-openjdk-jdk11/omr/gc/base/ModronAssertions.h:33:0,
                 from /root/code/openj9-openjdk-jdk11/openj9/runtime/gc_base/ObjectAccessBarrier.hpp:41,
                 from /root/code/openj9-openjdk-jdk11/openj9/runtime/gc_base/FinalizeListManager.hpp:38,
                 from /root/code/openj9-openjdk-jdk11/openj9/runtime/gc_verbose_java/VerboseHandlerJava.cpp:36:
/root/code/openj9-openjdk-jdk11/build/linux-x86_64-normal-server-release/vm/runtime/omr/gc/ut_j9mm.h:30:0: error: "UT_MODULE_LOADED" redefined [-Werror]
```
And move app around does work! Thanks.

File: gc/verbose/VerboseHandlerOutput.cpp
Comment (RSalman) [inline gc/verbose/VerboseHandlerOutput.cpp:1049]: Going to manager and getting handler is redundant, `getTagTemplate` can be called directly here

File: gc/verbose/VerboseHandlerOutput.cpp
Comment (RSalman) [inline gc/verbose/VerboseHandlerOutput.cpp:262]: Is this needed, can it be deleted now that we have `outputInitializedRegion`? The call sites can directly call `outputInitializedRegion`. I think it only ever called through `handleInitialized` and `openFile`

**Review by RSalman — CHANGES_REQUESTED** [review]

File: gc/verbose/VerboseHandlerOutput.cpp
Comment (RSalman) [reply gc/verbose/VerboseHandlerOutput.cpp:262]: Similarly, `handleInitializedInnerStanzas ` can be deleted now that you have the `outputInitializedInnerStanzas` variant

**Review by RSalman — CHANGES_REQUESTED** [review]

File: gc/verbose/VerboseWriterFileLogging.hpp
Comment (RSalman) [inline gc/verbose/VerboseWriterFileLogging.hpp:73]: Default param here is false, but true in `VerboseWriterFileLoggingBuffered.hpp` and `VerboseWriterFileLoggingSynchronous.hpp `. Is this intended?

File: gc/verbose/VerboseWriterFileLoggingBuffered.cpp
Comment (RSalman) [inline gc/verbose/VerboseWriterFileLoggingBuffered.cpp:29]: Formatting: new line not needed

File: gc/verbose/VerboseWriterFileLoggingBuffered.cpp
Comment (RSalman) [inline gc/verbose/VerboseWriterFileLoggingBuffered.cpp:159]: Change to`/*` style + more specific, as explained in my other review comment

File: gc/verbose/VerboseWriterFileLoggingSynchronous.cpp
Comment (RSalman) [inline gc/verbose/VerboseWriterFileLoggingSynchronous.cpp:160]: See my comment in  `VerboseWriterFileLoggingBuffered.cpp`

**Review by RSalman — CHANGES_REQUESTED** [review]

**Review by RSalman — CHANGES_REQUESTED** [review]
Although these changes won't break OpenJ9, they're still incompatible. Inner stanzas won't be printed, a call to `outputInitializedInnerStanza` or to `writeVMargs` will result in this base class empty implementations being called.

File: gc/verbose/VerboseHandlerOutput.cpp
Comment (RSalman) [inline gc/verbose/VerboseHandlerOutput.cpp:267]: This can be removed from here. `InitializedTime` should only be set once, during startup of GC and it's handled in line 374. This time not actually used by this stanza, its just a marker of when GC is initialized and used to calculate interval between 2 of the same stanzas, when other stanzas are printed and there is 1 stanza, so this time is used.

Comment (RSalman) [reply gc/verbose/VerboseHandlerOutput.cpp:267]: You can Assert instead, Assert(_manager->getInitializedTime() != 0)

**Review by RSalman — CHANGES_REQUESTED** [review]

File: gc/verbose/VerboseHandlerOutput.cpp
Comment (RSalman) [inline gc/verbose/VerboseHandlerOutput.cpp:374]: `event->timestamp` instead of  `omrtime_hires_clock `

Comment (RSalman) [reply gc/verbose/VerboseHandlerOutput.cpp:374]: and then you can also remove `OMRPORT_ACCESS_FROM_OMRPORT(env->getPortLibrary());`

File: gc/verbose/VerboseWriterFileLoggingBuffered.cpp
Comment (RSalman) [inline gc/verbose/VerboseWriterFileLoggingBuffered.cpp:123]: Minor formatting: space before `{`

File: gc/verbose/VerboseWriterFileLoggingBuffered.cpp
Comment (RSalman) [inline gc/verbose/VerboseWriterFileLoggingBuffered.cpp:125]: space before `{`

File: gc/verbose/VerboseWriterFileLoggingSynchronous.cpp
Comment (RSalman) [inline gc/verbose/VerboseWriterFileLoggingSynchronous.cpp:124]: Formatting for this block, see comment for `VerboseWriterFileLoggingBuffered.cpp`

**Review by RSalman — CHANGES_REQUESTED** [review]
setInitalizedTime got mixed up + some formatting.

Also, you can go ahead with cleanup/TODO items in this PR. As Aleks mentioned, it's fine to have missing section of VGC while waiting for OpenJ9 merge.

**Comment by amicic** [comment]
What is the motivation for moving formatAndOutput() from Writer to Buffer? Why couldn't we just call _manager->getWriterChain()->handleInitialized() from openFile() ?

**Comment by RSalman** [comment]
We extracted code from handleInit to outputInit to generalize the logic. handleInit acted specifically on Writer chain's buffer, we changed things to act on any provided buffer. There is a level of abstraction with all the handle* methods,  they call Writer chains output method, which just outputs contents to its own buffer. When we call flush on writer chain, the buffer contents are outputted to all the writers. With the changes, handleInit will pass in the wirterchain buffer while openFile pass in a new buffer.

Using the existing handleInit for openFile causes a few complications. For one, when the writer chain gets flushed _(which would be done by handleInit)_, it outputs the buffer to all the writers, which is not correct, we only want it outputted to the file being opened. You can argue that we can just filter for the file writer when opening a new file and only output to that. But there's also another issue, when we're opening a file, we're in the middle of a flush from another  handle* method. That is, the flush realizes that a file needs to be opened before flushing contents of buffer. We can't simply just branch out to another handle method. Only way is to save and restore the writer chain buffer, this would lead to more messy implementation.

File: gc/include/omrmm.hdf
Comment (amicic) [inline gc/include/omrmm.hdf:209]: is this temporary (is there intention for this reverting back to plain J9HOOK_MM_OMR_INITIALIZED)?

if not, perhaps this could be called ..._V2 ?

**Comment by amicic** [comment]
@gza060625 please update Copyright year. In several files it's 2016 or so.

Also squash all the commits into 1.

File: gc/include/omrmm.hdf
Comment (RSalman) [inline gc/include/omrmm.hdf:209]: Why do we need to add an extra event, I'm missing something, but is this really needed?

**Review by RSalman — APPROVED** [review]
LGTM

**Review by amicic — APPROVED** [review]

**Review by RSalman — CHANGES_REQUESTED** [review]

**Review by RSalman — APPROVED** [review]
Approving assuming minor formatting/nitpicks are addressed + commits are squashed

**Comment by gza060625** [comment]
@fjeremic  and @vijaysun-omr please review and merge. Thanks

**Comment by fjeremic** [comment]
@gza060625 could we modify the commit description to conform to the Commit Guidelines [1]. I'll point out the relevant sections:

> The first line describes the change made. It is written in the imperative mood, and should say what happens when the patch is applied. Keep it short and simple. The first line should be less than 70 characters, where reasonable, and should be written in sentence case preferably not ending in a period. Leave a blank line between the first line and the message body.
> ...
> The body should be wrapped at 72 characters, where reasonable.
> ...
> Use the commit footer to place commit metadata. The footer is the last block of contiguous text in the message. It is separated from the body by one or more blank lines, and as such cannot contain any blank lines.

I've gone ahead and pretty-printed your commit. Ideally if you only have one commit it should match the PR title.

```
Print `initialized` stanza to each file

In order to achieve this, following changes have been made:

- Relocate `formatAndOutput(),formatAndOutputV()` to `MM_VerboseBuffer`
  from `MM_VerboseWriterChain`
- In `MM_VerboseHandlerOutput`:
   - Modified `writeVmArgs` to take a `MM_VerboseBuffer`
   - Factored out lines for printing `Initialized` stanza in `handleInitialized`
   - Created `outputInitializedStanza` dedicated to output `Initialized`
     stanza, which also takes a `MM_VerboseBuffer`
   - In `outputInitializedStanza`, replaced `handleInitializedInnerStanzas`
     with `outputInitializedInnerStanza`
- In `VerboseHandlerOutput`:
   - Modified `writeVmArgs` to takes a `MM_VerboseBuffer`
   - Implement virtual method `outputInitializedInnerStanza()`
- Implement `MM_VerboseManager::getVerboseHandlerOutput()`
- In `VerboseWriterChain`:
   - Delete `formatAndOutputV()`
   - Redirect `formatAndOutput()` to `MM_VerboseBuffer`
   - Implement `getBuffer()`
- In `MM_VerboseWriterFileLoggingBuffered` and `VerboseWriterFileLoggingSynchronous`:
   - Print `Initialized` stanza on the second file opens (First file is
     handled by `TRIGGER_J9HOOK_MM_OMR_INITIALIZED`)
- In `MM_VerboseWriterFileLogging::endOfCycle`, opens a file right
  after last file closes to make sure a new `Initialized` stanza is
  printed in `openFile(...)`.
- Changes to generalize handleIniaitzed to work on any provided buffer
  rather than writer chain specific buffer. This is done to bypass the
  writer chain with a new buffer during openFile to print stanza 
  specifically for the file being opened.

Signed-off-by: Enson Guo <enson.guo@ibm.com>
```

[1] https://github.com/eclipse/omr/blob/master/CONTRIBUTING.md#commit-guidelines

**Comment by gza060625** [comment]
@fjeremic Thanks for pointing it out. I have updated the commit description.

**Comment by fjeremic** [comment]
@genie-omr build all

**Comment by fjeremic** [comment]
@gza060625 see failure on x86 compressedrefs. Looks like a legitimate issue.

**Comment by amicic** [comment]
Indeed, GCExtensions and ObjectAccessBarrier are OpenJ9 specific. I see they are used now to directly obtain info about CR (shift etc), but it used it to be passed indirectly through a hook. This is not trivial, let me think....

File: gc/verbose/VerboseHandlerOutput.cpp
Comment (amicic) [inline gc/verbose/VerboseHandlerOutput.cpp:343]: could we not pass eventData down to outputInitializedStanza and get relevant info for printing from the event rather than accessing GCExt?

File: gc/verbose/VerboseHandlerOutput.cpp
Comment (amicic) [inline gc/verbose/VerboseHandlerOutput.cpp:261]: CR shift is the only thing from accessBarrier. We can get it from omrVM->_compressedPointersShift, and thus avoid accessing accessBarrier and MM_GCExtensions that are OpenJ9 specific

**Comment by amicic** [comment]
@fjeremic CR failure has been address. Let's try the builds again.

**Comment by fjeremic** [comment]
@genie-omr build all

**Pull request merged**
